File: 2.AddTwoNumbers/Solution.cpp
#include <bits/stdc++.h>
using namespace std;
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
class Solution
{
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2){
        ListNode *l3 = new ListNode(), *p = l3;
        int sum = 0;
        while(l1 or l2){
            if(l1){
                sum += l1->val;
                l1 = l1->next;
            }
            if (l2){
                sum += l2->val;
                l2 = l2->next;
            }
            p->val = sum%10;
            sum /= 10;
            p->next = new ListNode();
            p = p->next;
        }
        return l3;
    }
};

File: 2.AddTwoNumbers/Solution.java
public class Solution {
    class ListNode {
        int val;
        ListNode next;
    
        ListNode() {
        }
    
        ListNode(int val) {
            this.val = val;
        }
    
        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode l3 = new ListNode(), p = l3;
        int sum = 0;
        while (l1 != null || l2 != null) {
            if (l1 != null) {
                sum += l1.val;
                l1 = l1.next;
            }
            if (l2 != null) {
                sum += l2.val;
                l2 = l2.next;
            }
            l3.val = sum % 10;
            sum /= 10;
            if (l1 != null || l2 != null || sum == 1) {
                l3.next = new ListNode();
            }
            l3 = l3.next;
        }
        if (sum) {
            l3.val = 1;
        }
        return p;
    }
}

